c/tskit/tables.c

```diff
@@ -7799,6 +7799,317 @@ tsk_table_collection_clear(tsk_table_collection_t *self)
+        node_map[source->node[row]], population_map[source->source[row]],
+        population_map[source->dest[row]], source->time[row],
+        source->metadata + source->metadata_offset[row], metadata_length);
+    if (ret < 0) {
+        goto out;
+    }
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_table_collection_subset(
+    tsk_table_collection_t *self, tsk_id_t *nodes, size_t num_nodes)
+{
+    int ret = 0;
+    tsk_id_t k, n, m, ind, pop, mut, site;
+    tsk_id_t *node_map, *individual_map, *population_map, *site_map, *mutation_map;
+    tsk_table_collection_t tables;
+
+    node_map = calloc(self->nodes.num_rows, sizeof(tsk_id_t));
+    individual_map = calloc(self->individuals.num_rows, sizeof(tsk_id_t));
+    population_map = calloc(self->populations.num_rows, sizeof(tsk_id_t));
+    site_map = calloc(self->sites.num_rows, sizeof(tsk_id_t));
+    mutation_map = calloc(self->mutations.num_rows, sizeof(tsk_id_t));
+    if (node_map == NULL || individual_map == NULL || population_map == NULL
+        || site_map == NULL || mutation_map == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    memset(node_map, 0xff, self->nodes.num_rows * sizeof(tsk_id_t));
+    memset(individual_map, 0xff, self->individuals.num_rows * sizeof(tsk_id_t));
+    memset(population_map, 0xff, self->populations.num_rows * sizeof(tsk_id_t));
+    memset(site_map, 0xff, self->sites.num_rows * sizeof(tsk_id_t));
+    memset(mutation_map, 0xff, self->mutations.num_rows * sizeof(tsk_id_t));
+
+    ret = tsk_table_collection_copy(self, &tables, 0);
+    if (ret != 0) {
+        goto out;
+    }
+    ret = tsk_table_collection_clear(self);
+    if (ret != 0) {
+        goto out;
+    }
+
+    // nodes, individuals, populations
+    for (k = 0; k < (tsk_id_t) num_nodes; k++) {
+        n = nodes[k];
+        node_map[n] = k;
+        ind = tables.nodes.individual[n];
+        if (ind >= 0) {
+            if (individual_map[ind] < 0) {
+                ret = individual_table_copy_row(
+                    &tables.individuals, &self->individuals, ind);
+                if (ret < 0) {
+                    goto out;
+                }
+                individual_map[ind] = ret;
+            }
+        }
+        pop = tables.nodes.population[n];
+        if (pop >= 0) {
+            if (population_map[pop] < 0) {
+                ret = population_table_copy_row(
+                    &tables.populations, &self->populations, pop);
+                if (ret < 0) {
+                    goto out;
+                }
+                population_map[tables.nodes.population[n]] = ret;
+            }
+        }
+        ret = node_table_copy_row(
+            &tables.nodes, &self->nodes, n, population_map, individual_map);
+        if (ret < 0) {
+            goto out;
+        }
+    }
+
+    // edges
+    for (k = 0; k < (tsk_id_t) tables.edges.num_rows; k++) {
+        m = node_map[tables.edges.parent[k]];
```

Comment:
I'm not sure we're getting much from the ``copy_row`` function here. What about
```c
edge_table_get_row(&tables.edges, k, &edge);
new_parent = node_map[edge.parent];
new_child = node_map[edge.child];
if (new_parent != TSK_NULL and new_child != TSK_NULL) {
    ret = edge_table_add_row(&self->edges,
          edge->left, edge->right, new_parent, new_child, edge->metadata, edge->metadata_length);
    if (ret < 0) {
        goto out;
    }
```

This actually makes life quite a bit simpler, because we don't need to test the ``edge_table_copy_row`` function now.